erpnext/controllers/accounts_controller.py

```diff
@@ -16,995 +17,1030 @@
+            return self.disable_rounded_total
+        else:
+            return frappe.db.get_single_value("Global Defaults", "disable_rounded_total")
 
 @frappe.whitelist()
 def get_tax_rate(account_head):
-	return frappe.db.get_value("Account", account_head, ["tax_rate", "account_name"], as_dict=True)
+    return frappe.db.get_value("Account", account_head, ["tax_rate", "account_name"], as_dict=True)
 
 @frappe.whitelist()
 def get_default_taxes_and_charges(master_doctype, tax_template=None, company=None):
-	if not company: return {}
+    if not company: return {}
 
-	if tax_template and company:
-		tax_template_company = frappe.db.get_value(master_doctype, tax_template, "company")
-		if tax_template_company == company:
-			return
+    if tax_template and company:
+        tax_template_company = frappe.db.get_value(master_doctype, tax_template, "company")
+        if tax_template_company == company:
+            return
 
-	default_tax = frappe.db.get_value(master_doctype, {"is_default": 1, "company": company})
+    default_tax = frappe.db.get_value(master_doctype, {"is_default": 1, "company": company})
 
-	return {
-		'taxes_and_charges': default_tax,
-		'taxes': get_taxes_and_charges(master_doctype, default_tax)
-	}
+    return {
+        'taxes_and_charges': default_tax,
+        'taxes': get_taxes_and_charges(master_doctype, default_tax)
+    }
 
 @frappe.whitelist()
 def get_taxes_and_charges(master_doctype, master_name):
-	if not master_name:
-		return
-	from frappe.model import default_fields
-	tax_master = frappe.get_doc(master_doctype, master_name)
+    if not master_name:
+        return
+    from frappe.model import default_fields
+    tax_master = frappe.get_doc(master_doctype, master_name)
 
-	taxes_and_charges = []
-	for i, tax in enumerate(tax_master.get("taxes")):
-		tax = tax.as_dict()
+    taxes_and_charges = []
+    for i, tax in enumerate(tax_master.get("taxes")):
+        tax = tax.as_dict()
 
-		for fieldname in default_fields:
-			if fieldname in tax:
-				del tax[fieldname]
+        for fieldname in default_fields:
+            if fieldname in tax:
+                del tax[fieldname]
 
-		taxes_and_charges.append(tax)
+        taxes_and_charges.append(tax)
 
-	return taxes_and_charges
+    return taxes_and_charges
 
 def validate_conversion_rate(currency, conversion_rate, conversion_rate_label, company):
-	"""common validation for currency and price list currency"""
+    """common validation for currency and price list currency"""
 
-	company_currency = frappe.db.get_value("Company", company, "default_currency", cache=True)
+    company_currency = frappe.db.get_value("Company", company, "default_currency", cache=True)
 
-	if not conversion_rate:
-		throw(_("{0} is mandatory. Maybe Currency Exchange record is not created for {1} to {2}.").format(
-			conversion_rate_label, currency, company_currency))
+    if not conversion_rate:
+        throw(_("{0} is mandatory. Maybe Cur
```

Comment:
**Purchase Order**
Following functions from validate should also be called.
1. Validate min_order_qty
https://github.com/frappe/erpnext/blob/develop/erpnext/buying/doctype/purchase_order/purchase_order.py#L59
2. Set raw_materials_supplied table, as it's depends on qty
https://github.com/frappe/erpnext/blob/develop/erpnext/buying/doctype/purchase_order/purchase_order.py#L61
3. Set status
https://github.com/frappe/erpnext/blob/develop/erpnext/buying/doctype/purchase_order/purchase_order.py#L47

Also, I think most of the functions called in `on_submit` should be called while updating qty / rate.
- Update last pur rate
- update status updater if PO is created against SO
- Update previous doc's status
- Update requested qty in bin
- Update ordered qty in bin
- update reserved qty for subcontract
- Update blanket order
- Validate budget
- Validate approving authority
https://github.com/frappe/erpnext/blob/develop/erpnext/buying/doctype/purchase_order/purchase_order.py#L205